File: src/C3.Catalogue/CassetteModels/CassetteModelOperationResult.vb
Namespace CassetteModels

    Public Enum CassetteModelFailure
        None = 0
        BrandRequired
        BrandNotFound
        InvalidType
        ModelNameRequired
        InvalidCode
        DuplicateIdentifier
        NotFound
        ReferencedByTape
        StorageFailure
    End Enum

    Public NotInheritable Class CassetteModelOperationResult

        Private Sub New()
        End Sub

        Public Property IsSuccess As Boolean
        Public Property Model As CassetteModel
        Public Property Failure As CassetteModelFailure
        Public Property Message As String

        Public Shared Function Success(value As CassetteModel) As CassetteModelOperationResult
            Return New CassetteModelOperationResult() With {
                .IsSuccess = True,
                .Model = value,
                .Failure = CassetteModelFailure.None,
                .Message = String.Empty
            }
        End Function

        Public Shared Function Failed(
                failure As CassetteModelFailure,
                message As String) As CassetteModelOperationResult

            Return New CassetteModelOperationResult() With {
                .IsSuccess = False,
                .Failure = failure,
                .Message = message
            }
        End Function

    End Class

End Namespace

File: src/C3.Catalogue/CassetteModels/CassetteModelOperationResult.cs
namespace C3.Catalogue.CassetteModels
{
    public enum CassetteModelFailure
    {
        None = 0,
        BrandRequired,
        BrandNotFound,
        InvalidType,
        ModelNameRequired,
        InvalidCode,
        DuplicateIdentifier,
        NotFound,
        ReferencedByTape,
        StorageFailure
    }

    public sealed class CassetteModelOperationResult
    {
        private CassetteModelOperationResult()
        {
        }

        public bool IsSuccess { get; set; }

        public CassetteModel Model { get; set; }

        public CassetteModelFailure Failure { get; set; }

        public string Message { get; set; }

        public static CassetteModelOperationResult Success(CassetteModel value)
        {
            return new CassetteModelOperationResult
            {
                IsSuccess = true,
                Model = value,
                Failure = CassetteModelFailure.None,
                Message = string.Empty
            };
        }

        public static CassetteModelOperationResult Failed(
            CassetteModelFailure failure,
            string message)
        {
            return new CassetteModelOperationResult
            {
                IsSuccess = false,
                Failure = failure,
                Message = message
            };
        }
    }
}
